source/mqtt/MqttConnection.cpp

```diff
@@ -22,6 +24,22 @@ namespace Aws
                 }
             }
 
+            std::shared_ptr<MqttConnection> MqttConnection::NewConnectionFromMqtt5Client(
+                std::shared_ptr<Crt::Mqtt5::Mqtt5Client> mqtt5client) noexcept
+            {
+                if (!mqtt5client || !*mqtt5client || !mqtt5client->m_client_core)
+                {
+                    AWS_LOGF_ERROR(AWS_LS_MQTT5_CLIENT, "Failed to create mqtt3 connection: Mqtt5 Client is invalid.");
+                    return nullptr;
+                }
+
+                /* Creating a MqttConnection would need access the underlying c mqtt5 client. The c mqtt5 client
+                ** should only be managed by the Mqtt5ClientCore to make sure it is not terminated. Use the client
+                ** core to create a new connection.
+                */
+                return mqtt5client->m_client_core->NewConnection(mqtt5client->m_mqtt5to3AdapterOptions.get());
+            }
+
```

Comment:
The creation process became unnecessary complicated: `MqttConnection` accesses a private member of `Mqtt5Client`, then it calls a private method of `Mqtt5ClientCore`, then that private method calls a private method of `MqttConnection`. So, basically the chain of accesses/calls looks like the following:
```
MqttConnection::createConn -> Mqtt5Client::core -> Mqtt5ClientCore::createConn -> MqttConnection::actualCreateConn
```

All the `MqttConnection::NewConnectionFromMqtt5Client` needs from `Mqtt5Client` is  an `MqttConnectionOptions` object. So one possible way is to add a method `Mqtt5Client::getMqttConnectionOptions`, private or public. It'll make the flow easier to understand.

Reply:
The underlying `aws_mqtt5_client` is also required. So, `Mqtt5Client` should expose it to `MqttConnection` too. But still, I think it's better than `MqttConnection` and `Mqtt5Client/Core` being friends to each other and using each other's private members.

Reply:
I was worried that if `MqttConnection` could directly access the `aws_mqtt5_client`, we would need make sure `MqttConnection` acquire a reference of `aws_mqtt5_client` so that the c client would not be released during the creation process. If that is the case, I think it would be cleaner that we always have the `Mqtt5ClientCore` to do any operations on `aws_mqtt5_client` as it already acquire a reference of the c client. 
However, thinking of it, if we passed the Mqtt5Client shared_ptr by value here, it should keep the Mqtt5Client alive and keep `Mqtt5ClientCore` and the `aws_mqtt5_client` reference as well. 
It should be good to directly access the `aws_mqtt5_client`, feel free to correct me if I was wrong. 
I will update the code.